commcare_connect/opportunity/tasks.py

```diff
@@ -73,3 +80,44 @@ def generate_payment_export(opportunity_id: int, export_format: str):
         content = content.encode()
     default_storage.save(export_tmp_name, ContentFile(content))
     return export_tmp_name
+
+
+@celery_app.task()
+def send_notification_inactive_users(opportunity_id: int):
+    opportunity = Opportunity.objects.get(opportunity_id=opportunity_id)
+    opportunity_accesses = OpportunityAccess.objects.filter(opportunity_id=opportunity_id)
+    for access in opportunity_accesses:
+        last_user_learn_module = (
+            CompletedModule.objects.filter(user=access.user, opportunity_id=opportunity_id).order_by("date").last()
+        )
+        has_started_deliver = OpportunityClaim.objects.exists(access=access)
+        if (
+            last_user_learn_module.date <= datetime.date.today() - datetime.timedelta(days=3)
+            and not has_started_deliver
+        ):
+            message = Message(
+                usernames=[access.user.username],
+                title=gettext(f"Resume your learning journey for {opportunity.name}"),
+                body=gettext(
+                    f"You have not completed your learning for {opportunity.name}."
+                    "Please complete the learning modules to start delivering visits."
+                ),
+            )
+            send_message(message)
+
+        last_user_deliver_visit = (
+            UserVisit.objects.filter(user=access.user, opportunity_id=opportunity_id).order_by("visit_date").last()
+        )
+        if (
+            last_user_deliver_visit.visit_date <= datetime.date.today() - datetime.timedelta(days=2)
+            and has_started_deliver
+        ):
+            message = Message(
+                usernames=[access.user.username],
+                title=gettext(f"Resume your job for {opportunity.name}"),
+                body=gettext(
+                    f"You have not completed your delivery visits for {opportunity.name}."
+                    "Please complete all the deliver visits to avail the payment."
+                ),
+            )
+            send_message(message)
```

Comment:
It would be nice to extract functions out of this to make it easier to follow, one for learn messages and one for deliver messages:

```
if has_started_delivery:
   send_deliver_messages(access)
else:
   send_learn_messages(access)
```

Reply:
https://github.com/dimagi/commcare-connect/pull/181/commits/717677822e616548efa8422b05a2c88349cc96fd